plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryTypeManager.java

```diff
@@ -368,19 +383,84 @@ private Optional<ColumnMapping> convertToTrinoType(Field field)
             case JSON:
                 return Optional.of(new ColumnMapping(jsonType, false));
             case STRUCT:
-                // create the row
-                FieldList subTypes = field.getSubFields();
-                checkArgument(!subTypes.isEmpty(), "a record or struct must have sub-fields");
-                List<RowType.Field> fields = subTypes.stream()
-                        .map(subField -> toRawTypeField(subField.getName(), subField))
-                        .collect(toImmutableList());
-                RowType rowType = RowType.from(fields);
-                return Optional.of(new ColumnMapping(rowType, false));
+                return Optional.of(new ColumnMapping(createRowType(field), false));
             default:
                 return Optional.empty();
         }
     }
 
+    private RowType createRowType(Field field)
+    {
+        FieldList subTypes = field.getSubFields();
+        checkArgument(!subTypes.isEmpty(), "a record or struct must have sub-fields");
+        List<RowType.Field> fields = subTypes.stream()
+                .map(subField -> toRawTypeField(subField.getName(), subField))
+                .collect(toImmutableList());
+        return RowType.from(fields);
+    }
+
+    public List<ColumnMetadata> convertToTrinoType(List<String> names, List<String> types)
+    {
+        return convertToTrinoType(names, types, Optional::empty);
+    }
+
+    public List<ColumnMetadata> convertToTrinoType(List<String> names, List<String> types, Supplier<Optional<TableInfo>> tableSupplier)
+    {
+        checkArgument(names.size() == types.size(), "Mismatched column names and types");
+
+        ImmutableList.Builder<ColumnMetadata> columns = ImmutableList.builder();
+        for (int i = 0; i < names.size(); i++) {
+            String name = names.get(i);
+            TypeSignature typeSignature;
+            try {
+                TypeInfo typeInfo = getTypeInfoFromTypeString(types.get(i));
+                typeSignature = toTypeSignature(typeInfo);
+            }
+            catch (UnsupportedTypeException e) {
+                Optional<TableInfo> table = tableSupplier.get();
```

Comment:
This logic may call `BigQueryClient#getTable` method per STRUCT column in a table. It should be only once per table.

Reply:
I memoized the supplier to also avoid calling it for table without any STRUCT type.